internal/pkg/config/watcher/watcher.go

```diff
@@ -98,14 +102,19 @@ func (c *K8sConfigMapWatcher) Watch(ctx context.Context, notifyMe chan<- interfa
 	watcher, err := c.client.ConfigMaps(c.Namespace).Watch(metav1.ListOptions{
 		LabelSelector: mapStringStringToLabelSelector(c.ConfigMapLabels),
 	})
+	defer watcher.Stop()
 	if err != nil {
 		return fmt.Errorf("unable to create watcher (possible serviceaccount RBAC/ACL failure?): %s", err.Error())
 	}
-	ch := watcher.ResultChan()
-	var e watch.Event
 	for {
 		select {
-		case e = <-ch:
+		case e, ok := <-watcher.ResultChan():
+			// channel may closed caused by HTTP timeout, should restart watcher
+			// detail at https://github.com/kubernetes/client-go/issues/334
+			if !ok {
```

Comment:
this is awesome! do you think it would also be useful to catch the following? I noticed the [code you linked](https://github.com/kubernetes/kubernetes/blob/32b37f5a85e8f2987f8367e611147cd4d4dfa4f9/staging/src/k8s.io/client-go/tools/cache/reflector.go#L319-L340) previously also has that.
```
import (
  "k8s.io/apimachinery/pkg/watch"
  apierrs "k8s.io/apimachinery/pkg/api/errors"
)
....
if e.Type == watch.Error {
  return apierrs.FromObject(e.Object)
}
```

Reply:
I think `apierrors.FromObject` just construct specify api-server Error from Object, if we catch that Error, we also should restart watcher.

Reply:
Hmm, I think this type of failure (`e.Type == watch.Error`) should _not_ restart the watcher. This error is not a known condition, and I am not comfortable saying that we know how to remediate it. So, deciding to restart the watcher here is 

IMO, we should add this:
```
if e.Type == watch.Error {
  return apierrs.FromObject(e.Object)
}
```
and have no extra logic to handle this condition in `cmd/main.go`. These errors should be fatal, because we do not know what happened, or how to handle them.

Reply:
Oops, sorry! I didnt realize the latest commit already included this! My bad :)